### ISSUE TITLE
ENH: Marray meta-thread

### ISSUE THREAD

**Issue by crusaderky** [scipy/scipy#22883]
### Is your feature request related to a problem? Please describe.

Currently, MArray is tested in scipy on a ad-hoc basis, outside of the `xp` pattern.
This leads to problems; namely I have observed on many cases tests in `stats` fail because the encroach on a use case in `special` that is only triggered by marray, and `scipy.special.tests` itself doesn't test marray. It also makes it complicated to scale the testing up to multiple wrapped backends.

Responding to https://github.com/scipy/scipy/pull/22686#pullrequestreview-2788220818:

I think we should
- treat `marray.numpy`, `marray.jax.numpy`, etc. as regular backends as a basis
- in marray, have a small tweak that special-cases array-api-compat when generating namespace names, so that you can read `marray.dask.array` and not `marray.scipy._lib.array_api_compat.dask.array`.
- add ad-hoc logic in `conftest` which treats `@skip_xp_backends("marray")` to match all backends that start with `marray.`
- also add ad-hoc logic which extends the skips, e.g. anything that skips `jax.numpy` also automatically skips `marray.jax.numpy`
- same for exceptions, although in many cases you'll have to explicitly disable it again, e.g.
```python
@skip_xp_backends(np_only=True, exceptions=["cupy"])  # enables marray.cupy too
@skip_xp_backends("marray", reason="No recursive delegation")
```
- While we're there, we should make the changes to conftest generic, so that they can be reused e.g. for pint-arrays.
- In `array_api_compat` and `array_api_extra`, add explicit support for marray otherwise things like `is_writeable_array`, `is_lazy_array`, and `xpx.at` will not work.
- In `xp_capabilities`, add marray as one/two extra columns to the table instead of 5 extra rows.

> Do we need to specify combinations of backend and device separately?

Probably not in unit documentation, most likely yes in testing for the sake of safety.
In documentation there could be a single extra MArray column with `(WARN) CPU only` if the wrapped backend works on both CPU and GPU but MArray doesn't.

> Since `marray` support will initially only be in `stats`, I think it's fine for _no_ tests to be performed with `marray` by default and for such a column to render only if `marray_backends` is specified.

This sounds like a whole lot of unnecessary complication. Once you have the array-api-compat and array-api-extras infrastructure in place, I expect everything to pass with the notable exception of functions that wrap around `xpx.lazy_apply` or more crudely call `xp.asarray(f(np.asarray(x))`.

> tests like `xp_assert_equal(stats.skew(xp.asarray([]), xp.asarray(xp.nan))` will fail because MArray would produce a masked value rather than a NaN.

This fundamentally stems from MArray's concretization functions (`__asarray__`, `__dlpack__`, `__buffer__`). returning unmasked results. IMHO they should raise an exception `if is_lazy_array(x.mask) or xp.any(x.mask)`. For float dtypes, you may want to consider swapping masked for NaN in the returned arrays instead (I don't have a strong opinion re. the correctness of such an operation).
Downstream of that, `xp_assert_equal` _and anything else that does want to cast MArrays back to numpy_ needs special treatment.

**Comment by lucascolley** [comment]
the bullet point plan sounds good to me from a high-level

**Comment by lucascolley** [comment]
@mdhaber did you have any opinions on Guido's plan?

**Comment by mdhaber** [comment]
I thought support for MArray would always be the exception rather than the rule. I mean, we could test that MArrays _with no mask_ work with all array API functions, but there is no reason to expect them to do the "right" thing if there are masked elements. It typically requires custom tests. So I actually started to go in the other dirrection - rather than skipping MArray in every function but a select few in stats (and maybe special), mark/document individual functions as supporting MArray, and write tests for all those functions. Currently, passing `marray=True` in `xp_capabilities` is what makes MArray support "public" by adding a note in the documentation.

There should be pretty strong property-based tests for all marked functions in [`test_marray.py`](https://github.com/scipy/scipy/blob/v1.18.0/scipy/stats/tests/test_marray.py). I think I've been pretty careful (except for testing all-masked slices - see gh-24876). But certainly it would be better to have some machinery that _ensures_ that those functions are tested if they are marked.  

And yes, I do think it would be nice to run separate tests with `scipy.special` rather than just doing it via `scipy.stats` functions. Very generic, property-based tests can be written for those since since essentially all functions in `special` are elementwise, so there is no interaction between masked and unmasked elements. Basically, just test that for unmasked elements, the `data` of the MArray matches the result with a regular array, and the `mask` of the marray is the `logical_or` (`|`) of all the input arrays. (No, unlike `np.ma`, output NaNs should NOT be masked!)
